File: java/validator-jackson/src/main/java/org/jsonstructure/jackson/validator/Compose.java
package org.jsonstructure.jackson.validator;

import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import org.jsonstructure.jackson.validator.error.CompositeError;
import org.jsonstructure.jackson.validator.error.ValidationError;
import org.jsonstructure.jackson.validator.loanword.Slice;

import static org.jsonstructure.jackson.validator.error.ValidationError.errorAt;

public class Compose {

    public static final String COMPOSE_SYMBOL = "\u0ADD";

    private static ValidationError mapMerge(ObjectNode dst, ObjectNode src, Slice<String> scope) {
        CompositeError errors = new CompositeError();
        Iterator<Map.Entry<String, JsonNode>> iterator = src.fields();
        while (iterator.hasNext()) {
            Map.Entry<String, JsonNode> entry = iterator.next();
            Slice<String> newScope = scope.append(entry.getKey());
            JsonNode dstValue = dst.get(entry.getKey());
            if (dstValue != null) {
                boolean srcMap = entry.getValue().isObject();
                boolean dstMap = dstValue.isObject();
                if ((srcMap && !dstMap) || (!srcMap && dstMap)) {
                    ValidationError err = errorAt("Attempted merge between map and non-map types", newScope);
                    errors.add(err);
                } else if (srcMap && dstMap) {
                    ValidationError err = mapMerge((ObjectNode) dstValue, (ObjectNode) entry.getValue(), newScope);
                    errors.add(err);
                } else {
                    dst.set(entry.getKey(), entry.getValue().deepCopy());
                }
            } else {
                dst.set(entry.getKey(), entry.getValue().deepCopy());
            }
        }
        return errors;
    }

    private static ValidationError doCompose(JsonNode target,
                                             ObjectNode types,
                                             ObjectNode fragments,
                                             Set<String> prev,
                                             Slice<String> scope) {
        CompositeError errors = new CompositeError();
        if (!target.isObject()) {
            return errorAt("definition is not a JSON object", scope);
        }
        ObjectNode object = (ObjectNode) target;
        Iterator<Map.Entry<String, JsonNode>> iterator = object.fields();
        while (iterator.hasNext()) {
            Map.Entry<String, JsonNode> entry = iterator.next();
            String key = entry.getKey();
            JsonNode value = entry.getValue();
            if (value.isObject()) {
                Slice<String> newScope = scope.append(key);
                errors.add(doCompose(value, types, fragments, prev, newScope));
            }
        }
        if (errors.size() > 0) {
            return errors;
        }
        JsonNode composeNode = object.get(COMPOSE_SYMBOL);
        if (composeNode == null) {
            return null;
        }
        if (!composeNode.isArray()) {
            return errorAt("'\\u0ADD' property must be a string array", scope);
        }
        ArrayNode defs = (ArrayNode) composeNode;
        ObjectNode dest = new ObjectNode(Jackson.NODE_FACTORY);
        Iterator<JsonNode> defIter = defs.elements();
        while (defIter.hasNext()) {
            JsonNode defNode = defIter.next();
            if (!defNode.isTextual()) {
                errors.add(errorAt("'\\u0ADD' property must be a string array", scope));
                continue;
            }
            String defName = defNode.textValue();
            if (prev.contains(defName)) {
                errors.add(errorAt("cycle detected with definitions " + prev.toString(), scope));
                continue;
            }
            ObjectNode def;
            JsonNode tDef = types.get(defName);
            JsonNode fDef = fragments.get(defName);
            if ((tDef == null) && (fDef == null)) {
                errors.add(errorAt("Unknown definition " + defName, scope));
                continue;
            }
            if ((tDef != null) && (fDef != null)) {
                errors.add(errorAt("Internal error duplicate definition " + defName, scope));
                continue;
            }
            if (tDef != null) {
                Slice<String> newScope = Slice.<String>create("types", defName);
                Set<String> next = new HashSet<>(prev);
                next.add(defName);
                ValidationError err = doCompose(tDef, types, fragments, next, newScope);
                errors.add(err);
                if (err != null) {
                    continue;
                }
                def = (ObjectNode) tDef;
            } else {
                Slice<String> newScope = Slice.<String>create("fragments", defName);
                Set<String> next = new HashSet<>(prev);
                next.add(defName);
                ValidationError err = doCompose(fDef, types, fragments, next, newScope);
                errors.add(err);
                if (err != null) {
                    continue;
                }
                def = (ObjectNode) fDef;
            }
            ValidationError err = mapMerge(dest, def, scope);
            errors.add(err);
        }
        ValidationError err = mapMerge(dest, object, scope);
        errors.add(err);
        if (errors.size() > 0) {
            return errors;
        }
        object.removeAll();
        object.setAll(dest);
        object.remove(COMPOSE_SYMBOL);
        return null;
    }

    public static ValidationError compose(JsonNode shellNode) {
        CompositeError errors = new CompositeError();
        if (!shellNode.isObject()) {
            return errorAt("root object must be a JSON object", Slice.empty());
        }
        ObjectNode shell = (ObjectNode) shellNode;
        JsonNode typesNode = shell.get("types");
        JsonNode fragmentsNode = shell.get("fragments");
        JsonNode mainNode = shell.get("main");
        if ((typesNode == null) || typesNode.isNull()) {
            typesNode = new ObjectNode(Jackson.NODE_FACTORY);
        }
        if ((fragmentsNode == null) || fragmentsNode.isNull()) {
            fragmentsNode = new ObjectNode(Jackson.NODE_FACTORY);
        }
        if (!typesNode.isObject()) {
            errors.add(errorAt("'types' property must be a JSON object", Slice.empty()));
        }
        if (!fragmentsNode.isObject()) {
            errors.add(errorAt("'fragments' property must be a JSON object", Slice.empty()));
        }
        if (errors.size() > 0) {
            return errors.simplify();
        }
        ObjectNode types = (ObjectNode) typesNode;
        ObjectNode fragments = (ObjectNode) fragmentsNode;
        removeNullValues(types);
        removeNullValues(fragments);
        Set<String> dupls = intersection(types, fragments);
        if (dupls.size() > 0) {
            return errorAt("Duplicate keys across 'types' and 'fragments': " + dupls, Slice.empty());
        }
        Iterator<Map.Entry<String, JsonNode>> iterator = fragments.fields();
        while (iterator.hasNext()) {
            Map.Entry<String, JsonNode> entry = iterator.next();
            Slice<String> scope = Slice.<String>create("fragments", entry.getKey());
            Set<String> prev = new HashSet<>();
            prev.add(entry.getKey());
            if (PrimitiveTypes.PRIMITIVE_TYPES.contains(entry.getKey())) {
                errors.add(errorAt("Cannot declare fragment with primitive type name", scope));
            } else {
                errors.add(doCompose(entry.getValue(), types, fragments, prev, scope));
            }
        }
        iterator = types.fields();
        while (iterator.hasNext()) {
            Map.Entry<String, JsonNode> entry = iterator.next();
            Slice<String> scope = Slice.<String>create("types", entry.getKey());
            Set<String> prev = new HashSet<>();
            prev.add(entry.getKey());
            if (PrimitiveTypes.PRIMITIVE_TYPES.contains(entry.getKey())) {
                errors.add(errorAt("Cannot declare type with primitive type name", scope));
            } else {
                errors.add(doCompose(entry.getValue(), types, fragments, prev, scope));
            }
        }
        if (mainNode != null) {
            Slice<String> scope = Slice.<String>create("main");
            Set<String> prev = new HashSet<>();
            errors.add(doCompose(mainNode, types, fragments, prev, scope));
        }
        return errors.simplify();
    }

    private static void removeNullValues(ObjectNode types) {
        Iterator<Map.Entry<String, JsonNode>> iterator = types.fields();
        while (iterator.hasNext()) {
            JsonNode next = iterator.next().getValue();
            if (next.isNull()) {
                iterator.remove();
            }
        }
    }

    private static Set<String> intersection(ObjectNode a, ObjectNode b) {
        Set<String> aNames = new HashSet<>();
        Set<String> bNames = new HashSet<>();
        a.fieldNames().forEachRemaining(aNames::add);
        b.fieldNames().forEachRemaining(bNames::add);
        aNames.retainAll(bNames);
        return aNames;
    }
}

File: golang-jsonstructure/compose.go
package jsonstructure

import (
	"errors"
	"fmt"

	"github.com/mohae/deepcopy"
	multierror "github.com/mspiegel/go-multierror"
)

var ComposeSymbol = "\u0ADD"

func mapMerge(dst map[string]interface{}, src map[string]interface{}, scope []string) error {
	var errs error
	for k, v := range src {
		newscope := append(scope, k)
		srcMap, srcOK := v.(map[string]interface{})
		dstValue := dst[k]
		if dstValue != nil {
			dstMap, dstOK := dstValue.(map[string]interface{})
			if (srcOK && !dstOK) || (!srcOK && dstOK) {
				err := errors.New("Attempted merge between map and non-map types")
				err = errorAt(err, newscope)
				errs = multierror.Append(errs, err)
			} else if srcOK && dstOK {
				err := mapMerge(dstMap, srcMap, newscope)
				errs = multierror.Append(errs, err)
			} else {
				dst[k] = deepcopy.Copy(v)
			}
		} else {
			dst[k] = deepcopy.Copy(v)
		}
	}
	return errs
}

func doCompose(target interface{},
	types map[string]interface{},
	fragments map[string]interface{},
	prev map[string]bool,
	scope []string) error {
	var err, errs error
	object, ok := target.(map[string]interface{})
	if !ok {
		err = errors.New("definition is not a JSON object")
		err = errorAt(err, scope)
		return err
	}
	for k, v := range object {
		if obj, ok2 := v.(map[string]interface{}); ok2 {
			newscope := append(scope, k)
			err = doCompose(obj, types, fragments, prev, newscope)
			errs = multierror.Append(errs, err)
		}
	}
	if errs != nil {
		return errs
	}
	c := object[ComposeSymbol]
	if c == nil {
		return nil
	}
	defs, ok3 := c.([]interface{})
	if !ok3 {
		err = errors.New("'\\u0ADD' property must be an string array")
		return errorAt(err, scope)
	}
	dest := make(map[string]interface{})
	for _, iDef := range defs {
		defName, ok := iDef.(string)
		if !ok {
			err = errors.New("'\\u0ADD' property must be an string array")
			err = errorAt(err, scope)
			errs = multierror.Append(errs, err)
			continue
		}
		if prev[defName] {
			keys := keysSet(prev)
			err = fmt.Errorf("cycle detected with definitions %v", keys)
			err = errorAt(err, scope)
			errs = multierror.Append(errs, err)
			continue
		}
		tDef := types[defName]
		fDef := fragments[defName]
		if (tDef == nil) && (fDef == nil) {
			err = fmt.Errorf("Unknown definition %s", defName)
			err = errorAt(err, scope)
			errs = multierror.Append(errs, err)
			continue
		}
		if (tDef != nil) && (fDef != nil) {
			// this should have been previously detected
			err = fmt.Errorf("Internal error duplicate definition %s", defName)
			err = errorAt(err, scope)
			errs = multierror.Append(errs, err)
			continue
		}
		var def map[string]interface{}
		if tDef != nil {
			next := cloneSet(prev)
			next[defName] = true
			newscope := []string{"types", defName}
			err = doCompose(tDef, types, fragments, next, newscope)
			errs = multierror.Append(errs, err)
			if err != nil {
				continue
			}
			def = tDef.(map[string]interface{})
		} else {
			next := cloneSet(prev)
			next[defName] = true
			newscope := []string{"fragments", defName}
			err = doCompose(fDef, types, fragments, next, newscope)
			errs = multierror.Append(errs, err)
			if err != nil {
				continue
			}
			def = fDef.(map[string]interface{})
		}
		err = mapMerge(dest, def, scope)
		errs = multierror.Append(errs, err)
	}
	err = mapMerge(dest, object, scope)
	errs = multierror.Append(errs, err)
	if errs != nil {
		return errs
	}
	for k := range object {
		delete(object, k)
	}
	for k, v := range dest {
		object[k] = v
	}
	delete(object, ComposeSymbol)
	return nil
}

func removeNilValues(m map[string]interface{}) {
	for k, v := range m {
		if v == nil {
			delete(m, k)
		}
	}
}

func Compose(shell map[string]interface{}) error {
	var typ, frag map[string]interface{}
	var ok bool
	var err, errs error
	t := shell["types"]
	f := shell["fragments"]
	m := shell["main"]
	if t == nil {
		t = make(map[string]interface{})
	}
	if f == nil {
		f = make(map[string]interface{})
	}
	typ, ok = t.(map[string]interface{})
	if !ok {
		err = errors.New("'types' property must be a JSON object")
		err = errorAt(err, nil)
		errs = multierror.Append(errs, err)
	}
	frag, ok = f.(map[string]interface{})
	if !ok {
		err = errors.New("'fragments' property must be a JSON object")
		err = errorAt(err, nil)
		errs = multierror.Append(errs, err)
	}
	if errs != nil {
		return errs
	}
	removeNilValues(typ)
	removeNilValues(frag)
	dupls := intersection(typ, frag)
	if len(dupls) > 0 {
		err = fmt.Errorf("Duplicate keys across 'types' and 'fragments': %v", dupls)
		err = errorAt(err, nil)
		return err
	}
	for k, v := range frag {
		prev := map[string]bool{k: true}
		scope := []string{"fragments", k}
		if PrimitiveTypes[k] {
			err = errors.New("Cannot declare fragment with primitive type name")
			err = errorAt(err, scope)
			errs = multierror.Append(errs, err)
		} else if v != nil {
			err = doCompose(v, typ, frag, prev, scope)
			errs = multierror.Append(errs, err)
		}
	}
	for k, v := range typ {
		prev := map[string]bool{k: true}
		scope := []string{"types", k}
		if PrimitiveTypes[k] {
			err = errors.New("Cannot declare type with primitive type name")
			err = errorAt(err, scope)
			errs = multierror.Append(errs, err)
		} else if v != nil {
			err = doCompose(v, typ, frag, prev, scope)
			errs = multierror.Append(errs, err)
		}
	}
	if m != nil {
		prev := map[string]bool{}
		scope := []string{"main"}
		err = doCompose(m, typ, frag, prev, scope)
		errs = multierror.Append(errs, err)
	}
	return errs
}

func intersection(m1 map[string]interface{}, m2 map[string]interface{}) []string {
	var iSect []string
	if len(m1) > len(m2) {
		tmp := m1
		m1 = m2
		m2 = tmp
	}
	for k := range m1 {
		if _, ok := m2[k]; ok {
			iSect = append(iSect, k)
		}
	}
	return iSect
}
